pkg/admission/webhook/generic_webhook.go

```diff
@@ -80,47 +88,56 @@ func (p *WebhookDispatcher) Dispatch(ctx context.Context, attr admission.Attribu
 	return p.dispatcher.Dispatch(ctx, attr, o, whAccessor)
 }
 
-func (p *WebhookDispatcher) getAPIBindingWorkspace(attr admission.Attributes, lc logicalcluster.LogicalCluster) (logicalcluster.LogicalCluster, bool) {
-	l, err := p.apiBinding.List(labels.Everything())
+func (p *WebhookDispatcher) getAPIBindingWorkspace(attr admission.Attributes, clusterName logicalcluster.LogicalCluster) (logicalcluster.LogicalCluster, bool, error) {
+	parentClusterName, hasParent := clusterName.Parent()
+	if !hasParent {
+		// APIBindings in root are not possible (they can only point to sibling workspaces).
+		return logicalcluster.New(""), false, nil
+	}
+
+	objs, err := p.apiBindingsIndexer.ByIndex(byWorkspaceIndex, clusterName.String())
 	if err != nil {
-		return logicalcluster.New(""), false
+		return logicalcluster.New(""), false, err
 	}
-	for _, apiBinding := range l {
-		if apiBinding.GetClusterName() != lc.String() {
-			continue
-		}
+	for _, obj := range objs {
+		apiBinding := obj.(*apisv1alpha1.APIBinding)
 		for _, br := range apiBinding.Status.BoundResources {
+			if apiBinding.Status.BoundAPIExport.Workspace == nil {
+				// this will never happen today. But as soon as we add other reference types (like exports), this log output will remind out of necessary work here.
+				klog.Errorf("APIBinding %s has no referenced workspace", clusterName, apiBinding.Name)
+				continue
+			}
 			if br.Group == attr.GetResource().Group && br.Resource == attr.GetResource().Resource {
-				p, hasParent := logicalcluster.From(apiBinding).Parent()
-				if !hasParent {
-					klog.V(3).Infof("unable to find parent for apibinding: cluster:%v resource: %v/%v", apiBinding.GetClusterName(), apiBinding.GetNamespace(), apiBinding.GetName())
-					return logicalcluster.New(""), false
-				}
-				return p.Join(apiBinding.Status.BoundAPIExport.Workspace.WorkspaceName), true
+				return parentClusterName.Join(apiBinding.Status.BoundAPIExport.Workspace.WorkspaceName), true, nil
 			}
 		}
 	}
-	return logicalcluster.New(""), false
+	return logicalcluster.New(""), false, nil
 }
 
 // In the future use a restricted list call
 func (p *WebhookDispatcher) restrictToLogicalCluster(hooks []webhook.WebhookAccessor, lc logicalcluster.LogicalCluster) []webhook.WebhookAccessor {
+	// TODO(sttts): this might not scale if there are many webhooks. This is called per request, and traverses all
```

Comment:
In the future, could we add an indexer for webhooks based on logical cluster like we have for api bindings?

Reply:
the `Source` interface of kube does not allow that without patching. I think it is not really worth it because there are never many webhooks. We can of course add our own source implementation and use another interface. That might give us the option to improve this. But would wait until we really hit this bottleneck. I don't expect it.